File: reverse-linked-list/flynn.cpp
/**
 * 풀이
 * - 생략, 주석 참고
 * 
 * Big O
 * - N: 주어진 링크드 리스트 `head`의 길이
 * 
 * - Time complexity: O(N)
 * - Space complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) return head;
        // example: a - b - c - d

        ListNode* root = new ListNode();
        root->next = head;
        // root - a - b - c - d

        ListNode* p = head;
        ListNode* q = p->next;
        // root - a - b - c - d
        //       (p) (q)

        while (q != nullptr) {
            p->next = q->next;
            // root - a - c - d
            //        b /
            q->next = root->next;
            // root - a - c - d
            //    b /
            root->next = q;
            // root - b - a - c - d
            //       (q) (p)
            q = p->next;
            // root - b - a - c - d
            //           (p) (q)
        }

        return root->next;
    }
};

File: reorder-list/flynn.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func reorderList(head *ListNode) {
	// find a middle node
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	// reverse the second part of the list
	var prev, curr *ListNode = nil, slow
	for curr != nil {
		tmp := curr.Next
		curr.Next = prev
		prev = curr
		curr = tmp
	}
	// merge two parts of the list
	for curr1, curr2 := head, prev; curr2.Next != nil; {
		tmp1, tmp2 := curr1.Next, curr2.Next
		curr1.Next, curr2.Next = curr2, curr1.Next
		curr1, curr2 = tmp1, tmp2
	}
}
